File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/64424e7d595a92e13387c139bd9abd06.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define MAX_N 100
#define MAX_K 10

typedef long long ll;

ll max(ll a, ll b) {
    return (a > b) ? a : b;
}

ll sum(int arr[], int size) {
    ll sum = 0;
    for (int i = 0; i < size; i++) {
        sum += arr[i];
    }
    return sum;
}

ll rec(int l, int r, int dpi, int k, ll dps[][MAX_K+1], ll sums[], int arrs[][MAX_K]) {
    ll ans = 0;
    int carr[MAX_K];
    memcpy(carr, arrs[l], sizeof(int) * MAX_K);
    ll dp[MAX_K+1];
    memcpy(dp, dps[dpi], sizeof(ll) * (MAX_K+1));

    if (l+1 == r) {
        ans = dp[k];
        ll sum = 0;
        for (int i = 0; i < k && i < MAX_K; i++) {
            sum += carr[i];
            ans = max(ans, dp[k-i-1] + sum);
        }
    } else {
        int m = (l+r)/2;

        memcpy(dps[dpi+1], dp, sizeof(ll) * (MAX_K+1));
        for (int i = l; i < m-1; i++) {
            ll v = sums[i];
            int k1 = sizeof(arrs[i]) / sizeof(int);
            for (int j = k; j >= k1; j--) {
                dp[j] = max(dp[j], dp[j-k1] + v);
            }
        }
        ans = max(ans, rec(m, r, dpi+1, k, dps, sums, arrs));

        memcpy(dps[dpi+1], dp, sizeof(ll) * (MAX_K+1));
        for (int i = m; i < r-1; i++) {
            ll v = sums[i];
            int k1 = sizeof(arrs[i]) / sizeof(int);
            for (int j = k; j >= k1; j--) {
                dp[j] = max(dp[j], dp[j-k1] + v);
            }
        }
        ans = max(ans, rec(l, m, dpi+1, k, dps, sums, arrs));
    }

    return ans;
}

void solve() {
    int n, k;
    scanf("%d %d", &n, &k);

    ll sums[MAX_N];
    int arrs[MAX_N][MAX_K];

    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        for (int j = 0; j < t; j++) {
            scanf("%d", &arrs[i][j]);
        }
        if (t > k) {
            t = k;
        }
        sums[i] = sum(arrs[i], t);
    }

    ll dps[13][MAX_K+1];
    memset(dps, 0, sizeof(ll) * 13 * (MAX_K+1));

    printf("%lld\n", rec(0, n, 0, k, dps, sums, arrs));
}

int main() {
    solve();
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/64424e7d595a92e13387c139bd9abd06.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type DSum struct{}

func (ds *DSum) solve(in *Input) int64 {
	n, k := in.ReadInt(), in.ReadInt()
	sums := make([]int64, n)
	arrs := make([][]int64, n)

	for i := 0; i < n; i++ {
		t := in.ReadInt()
		arrs[i] = make([]int64, t)
		for j := 0; j < t; j++ {
			arrs[i][j] = in.ReadInt64()
		}
		if t > k {
			arrs[i] = arrs[i][:k]
		}
		sums[i] = sum(arrs[i])
	}

	dps := make([][]int64, 13)
	for i := range dps {
		dps[i] = make([]int64, k+1)
	}

	return ds.rec(0, n, 0, k, sums, arrs, dps)
}

func (ds *DSum) rec(l, r, dpi, k int, sums []int64, arrs [][]int64, dps [][]int64) int64 {
	ans := int64(0)
	carr := arrs[l]
	dp := dps[dpi]

	if l+1 == r {
		ans = max(ans, dp[k])
		sum := int64(0)
		for i := 0; i < min(k, len(carr)); i++ {
			sum += carr[i]
			ans = max(ans, dp[k-i-1]+sum)
		}
	} else {
		m := (l + r) / 2
		ds.doCopy(dpi, dps)
		for i := l; i < m; i++ {
			ds.addToKnap(i, dpi, sums, arrs, dps)
		}
		ans = max(ans, ds.rec(m, r, dpi+1, k, sums, arrs, dps))
		ds.doCopy(dpi, dps)
		for i := m; i < r; i++ {
			ds.addToKnap(i, dpi, sums, arrs, dps)
		}
		ans = max(ans, ds.rec(l, m, dpi+1, k, sums, arrs, dps))
	}
	return ans
}

func (ds *DSum) doCopy(dpi int, dps [][]int64) {
	copy(dps[dpi+1], dps[dpi])
}

func (ds *DSum) addToKnap(i, dpi int, sums []int64, arrs [][]int64, dps [][]int64) {
	v := sums[i]
	k1 := len(arrs[i])
	for j := len(dps[dpi]) - 1; j >= k1; j-- {
		dps[dpi][j] = max(dps[dpi][j], dps[dpi][j-k1]+v)
	}
}

func sum(arr []int64) int64 {
	total := int64(0)
	for _, v := range arr {
		total += v
	}
	return total
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

type Input struct {
	data []string
	index int
}

func (in *Input) ReadInt() int {
	val := in.ReadString()
	res := 0
	for _, c := range val {
		res = res*10 + int(c-'0')
	}
	return res
}

func (in *Input) ReadInt64() int64 {
	val := in.ReadString()
	res := int64(0)
	for _, c := range val {
		res = res*10 + int64(c-'0')
	}
	return res
}

func (in *Input) ReadString() string {
	if in.index >= len(in.data) {
		return ""
	}
	val := in.data[in.index]
	in.index++
	return val
}

func main() {
	var n, k int
	fmt.Scan(&n, &k)
	in := &Input{data: make([]string, n)}
	for i := 0; i < n; i++ {
		var t int
		fmt.Scan(&t)
		in.data[i] = fmt.Sprint(t)
		for j := 0; j < t; j++ {
			var x int64
			fmt.Scan(&x)
			in.data = append(in.data, fmt.Sprint(x))
		}
	}
	solver := DSum{}
	fmt.Println(solver.solve(in))
}
